File: food-ordering-system/restaurant-service/src/main/java/mk/finki/restaurantservice/domain/MenuItem.java
package mk.finki.restaurantservice.domain;

import jakarta.persistence.*;
import lombok.*;
import java.math.BigDecimal;
import java.util.UUID;

@Entity
@Table(name = "menu_items")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor(access = AccessLevel.PRIVATE)
@Builder
public class MenuItem {
    @Id
    private UUID id;

    private String name;
    private BigDecimal price;

    @Enumerated(EnumType.STRING)
    private MenuItemStatus status;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "restaurant_id")
    private Restaurant restaurant;

    // Package-private constructor for the Restaurant aggregate root to use
    MenuItem(String name, BigDecimal price, Restaurant restaurant) {
        this.id = UUID.randomUUID();
        this.name = name;
        this.price = price;
        this.status = MenuItemStatus.AVAILABLE;
        this.restaurant = restaurant;
    }

    public void updateStatus(MenuItemStatus status) {
        this.status = status;
    }
}

File: restaurant-service/src/main/kotlin/mk/finki/restaurantservice/domain/restaurant/entities/MenuItem.kt
package mk.finki.restaurantservice.domain.restaurant.entities

import io.swagger.v3.oas.annotations.media.Schema
import jakarta.persistence.*
import mk.finki.restaurantservice.domain.restaurant.enums.MenuItemStatus
import mk.finki.restaurantservice.domain.restaurant.valueobjects.MenuItemId
import mk.finki.restaurantservice.domain.restaurant.valueobjects.Money
import java.util.UUID

@Entity
@Table(name = "menu_items")
class MenuItem internal constructor(
    @EmbeddedId
    @Schema(example = "123e4567-e89b-12d3-a456-426614174002")
    val id: MenuItemId,
    @Schema(example = "Margherita Pizza")
    var name: String,
    @Embedded
    var price: Money,
    @Enumerated(EnumType.STRING)
    @Schema(example = "AVAILABLE")
    var status: MenuItemStatus
) {
    @Column(name = "deleted")
    @Schema(example = "false")
    var deleted: Boolean = false
        private set

    fun discontinue() {
        this.deleted = true
        this.status = MenuItemStatus.DISCONTINUED
    }

    fun updatePrice(newPrice: Money) {
        this.price = newPrice
    }

    fun updateStatus(newStatus: MenuItemStatus) {
        if (!deleted) {
            this.status = newStatus
        }
    }
    
    // For JPA
    protected constructor() : this(MenuItemId(UUID.randomUUID()), "", Money.zero(), MenuItemStatus.AVAILABLE)
}
